akka-remote/src/main/scala/akka/remote/RemoteWatcher.scala

```diff
@@ -207,57 +194,70 @@ private[akka] class RemoteWatcher(
-        watching -= ((watchee, self))
-      }
-      checkLastUnwatchOfNode(watchee.path.address)
+    log.debug("Unwatching: [{} -> {}]", watcher.path, watchee.path)
+    watching.get(watchee) match {
+      case Some(watchers) if watchers == Set(watcher) ⇒
+        clearAllWatches(watchee)
+        checkLastUnwatchOfNode(watchee.path.address)
+      case Some(watchers) ⇒
+        watching += watchee → (watchers - watcher)
+      case None ⇒
     }
+  }
 
-  def logActorForDeprecationWarning(watchee: ActorRef): Unit = {
+  def clearAllWatches(watchee: InternalActorRef): Unit = {
+    // clean up self watch when no more watchers of this watchee
+    log.debug("Cleanup self watch of [{}]", watchee.path)
+    context unwatch watchee
+    watching -= watchee
+  }
+
+  def logActorForDeprecationWarning(watchee: InternalActorRef): Unit = {
     log.debug("actorFor is deprecated, and watching a remote ActorRef acquired with actorFor is not reliable: [{}]", watchee.path)
   }
 
-  def terminated(watchee: ActorRef, existenceConfirmed: Boolean, addressTerminated: Boolean): Unit = {
+  def terminated(watchee: InternalActorRef, existenceConfirmed: Boolean, addressTerminated: Boolean): Unit = {
     log.debug("Watchee terminated: [{}]", watchee.path)
 
-    // When watchee is stopped it sends DeathWatchNotification to the watcher and to this RemoteWatcher,
-    // which is also watching. Send extra DeathWatchNotification to the watcher in case the
-    // DeathWatchNotification message is only delivered to RemoteWatcher. Otherwise there is a risk that
-    // the monitoring is removed, subsequent node failure is not detected and the original watcher is
-    // never notified. This may occur for normal system shutdown of the watchee system when not all remote
-    // messages are flushed at shutdown.
-    watching --= watching collect {
-      case tuple @ (wee, wer: InternalActorRef) if wee == watchee ⇒
-        if (!addressTerminated && wer != self)
-          wer.sendSystemMessage(DeathWatchNotification(watchee, existenceConfirmed, addressTerminated))
-        tuple
-    }
-
+    // When watchee is stopped it sends DeathWatchNotification to this RemoteWatcher,
+    // which will propagate it to all watchers of this watchee.
+    // addressTerminated case is already handled by the watcher itself in DeathWatch trait
+    if (!addressTerminated)
+      for {
+        watchers ← watching.get(watchee)
+        watcher ← watchers
+        if watcher != self
+      } watcher.sendSystemMessage(DeathWatchNotification(watchee, existenceConfirmed, addressTerminated))
+
+    watching -= watchee
     checkLastUnwatchOfNode(watchee.path.address)
   }
 
   def checkLastUnwatchOfNode(watcheeAddress: Address): Unit = {
-    if (watchingNodes(watcheeAddress) && watching.forall {
-      case (wee, wer) ⇒ wee.path.address != watcheeAddress
-    }) {
+    if (watchingNodes(watcheeAddress) && watching.keys.forall(_.path.address != watcheeAddress)) {
```

Comment:
Doing this for every termination can be quite costly; can we keep track of this property otherwise?